packages/connect/src/connector/extension.ts

```diff
@@ -0,0 +1,133 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+
+import type {
+  ToApplication,
+  ToExtension,
+} from "@substrate/connect-extension-protocol"
+import type { Chain, JsonRpcCallback, SubstrateConnector } from "./types.js"
+import {
+  AlreadyDestroyedError,
+  CrashError,
+  JsonRpcDisabledError,
+} from "./errors.js"
+import { SupportedChains } from "../SupportedChains.js"
+
+type HeaderlessToExtensionGeneric<T extends ToExtension> = T extends {
+  origin: "substrate-connect-client"
+} & infer V
+  ? Omit<V, "chainId">
+  : unknown
+type HeaderlessToExtension = HeaderlessToExtensionGeneric<ToExtension>
+
+const listeners = new Map<string, (msg: ToApplication) => void>()
+window.addEventListener("message", ({ data }: MessageEvent<ToApplication>) => {
+  if (data?.origin !== "substrate-connect-extension") return
+  listeners.get(data.chainId)?.(data)
+})
+
+const listenToExtension = (
+  chainId: string,
+  callback: (msg: ToApplication) => void,
+): (() => void) => {
+  listeners.set(chainId, callback)
+  return () => {
+    listeners.delete(chainId)
+  }
+}
+
+function getRandomChainId(): string {
+  const arr = new BigUint64Array(2)
+  // It can only be used from the browser, so this is fine.
+  crypto.getRandomValues(arr)
+  const result = (arr[1] << BigInt(64)) | arr[0]
+  return result.toString(36)
+}
+
+export const getConnectorClient = (): SubstrateConnector => {
+  const chains = new Map<Chain, string>()
+
+  const internalAddChain = async (
+    isWellKnown: boolean,
+    chainSpecOrWellKnownName: string,
+    jsonRpcCallback?: JsonRpcCallback,
+    potentialRelayChainIds = [] as string[],
+  ): Promise<Chain> => {
+    const chainId = getRandomChainId()
+    const postToExtension = (msg: HeaderlessToExtension) => {
+      window.postMessage(
+        {
+          ...msg,
+          origin: "substrate-connect-client",
+          chainId,
+        },
+        "*",
+      )
+    }
+
+    await new Promise<void>((res, rej) => {
+      const stopListeningToExtension = listenToExtension(chainId, (msg) => {
+        stopListeningToExtension()
+        if (msg.type === "chain-ready") return res()
+        rej(new Error("There was an error creating the smoldot chain."))
+      })
+
+      postToExtension(
+        isWellKnown
+          ? {
+              type: "add-well-known-chain",
+              chainName: chainSpecOrWellKnownName,
```

Comment:
One thing that needs to be handled is: if the extension fails to initialize the well-known chain, we should fall back to a regular `add-chain` message.

Reply:
how would that work, though? we would take the chainSpecs from the known ones and try again as a normal `addChain` message? What does that accomplish, though?

Reply:
> we would take the chainSpecs from the known ones and try again as a normal addChain message?

Yes

> What does that accomplish, though?

The idea is to handle situations where the extension removes support for a well-known chain.
Basically, it's as if you asked the extension "do you support chain N?" and if the extension says no you send the chain spec, except that instead of asking the extension if it supports the chain we simply directly ask the extension to initialize the chain, and check if an error is returned.

Reply:
It makes sense... I will do that, thanks!